# DSN Bootcamp Qualification Hackathon 2026
## Phase 1: Data Audit

The objective of this phase is to understand the structure and quality
of the training and test datasets before performing EDA, feature
engineering, validation, or model training.

### Audit objectives

1. Dataset structure
2. Data types
3. Missing values
4. Cardinality
5. Duplicate observations
6. Train/test consistency
7. Target variable audit


In [ ]:
# Import and load data
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

train = pd.read_csv("../data/train.csv")
test = pd.read_csv("../data/test.csv")

print("Train shape:", train.shape)
print("Test shape :", test.shape)
print("Sum of Train duplicates", train.duplicated().sum())
train

Train shape: (6818, 13)
Test shape : (1705, 12)


,id,product_code,product_weight_kg,fat_content,shelf_visibility,product_category,product_price,store_code,store_age_years,store_size,store_location_tier,store_format,total_sales
0,row_00000,PRD-PRFP9S,14.252,Low Fat,0.0271,Frozen Foods,81.37,STORE-AGY,45,Large,Tier_3,Standard Supermarket,1764.98
1,row_00001,PRD-PXXK71,7.698,Low Fat,0.0720,HEALTH AND HYGIENE,42.05,STORE-YLW,35,Small,Tier_1,Standard Supermarket,342.13
2,row_00002,PRD-V5MOIJ,14.264,Regular,0.0421,Canned,41.35,STORE-89Z,33,Medium,Tier_1,Standard Supermarket,378.85
3,row_00003,PRD-UN5Z3J,NaN,Regular,0.0449,soft drinks,174.35,STORE-7WS,47,Medium,Tier_3,Flagship Hypermarket,5595.72
4,row_00004,PRD-6RDQYB,10.338,Regular,0.0120,meat,203.06,STORE-9RG,28,Small,Tier_2,Standard Supermarket,2375.36
...,...,...,...,...,...,...,...,...,...,...,...,...,...
6813,row_08518,PRD-V56VKT,9.201,Regular,0.2851,Fruits and Vegetables,136.57,STORE-JOR,34,NaN,Tier_3,Corner Shop,276.50
6814,row_08519,PRD-S1AWP0,15.770,Low Fat,0.1193,FROZEN FOODS,74.03,STORE-OYG,25,NaN,Tier_2,Standard Supermarket,1265.63
6815,row_08520,PRD-NCP2VI,17.665,Low Fat,0.0182,Health and Hygiene,234.87,STORE-DKU,30,NaN,Tier_2,Standard Supermarket,6254.55
6816,row_08521,PRD-K9LSG4,20.645,Low Fat,0.0567,snack foods,118.86,STORE-OYG,25,NaN,Tier_2,Standard Supermarket,1658.06


In [17]:
# Inspect Columns
print("Train Columns")

for i, col in enumerate(train.columns, start=1):
    print(f"{i:2}. {col}")


print("\nTest Columns")

for i, col in enumerate(test.columns, start=1):
    print(f"{i:2}. {col}")

Train Columns
 1. id
 2. product_code
 3. product_weight_kg
 4. fat_content
 5. shelf_visibility
 6. product_category
 7. product_price
 8. store_code
 9. store_age_years
10. store_size
11. store_location_tier
12. store_format
13. total_sales

Test Columns
 1. id
 2. product_code
 3. product_weight_kg
 4. fat_content
 5. shelf_visibility
 6. product_category
 7. product_price
 8. store_code
 9. store_age_years
10. store_size
11. store_location_tier
12. store_format


In [18]:
# Basic Structure
print("Train Info")
train.info()

print("\nTest Info")
test.info()

Train Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6818 entries, 0 to 6817
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   6818 non-null   object 
 1   product_code         6818 non-null   object 
 2   product_weight_kg    5593 non-null   float64
 3   fat_content          6818 non-null   object 
 4   shelf_visibility     6818 non-null   float64
 5   product_category     6818 non-null   object 
 6   product_price        6818 non-null   float64
 7   store_code           6818 non-null   object 
 8   store_age_years      6818 non-null   int64  
 9   store_size           4899 non-null   object 
 10  store_location_tier  6818 non-null   object 
 11  store_format         6818 non-null   object 
 12  total_sales          6818 non-null   float64
dtypes: float64(4), int64(1), object(8)
memory usage: 692.6+ KB

Test Info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1705 en

In [19]:
# Data Type Audit
dtype_audit = pd.DataFrame({
    "dtype": train.dtypes.astype(str),
    "missing": train.isna().sum(),
    "missing_pct": (train.isna().mean() * 100).round(2),
    "unique_values": train.nunique(dropna=True)
})

dtype_audit

,dtype,missing,missing_pct,unique_values
id,object,0,0.00,6818
product_code,object,0,0.00,1555
product_weight_kg,float64,1225,17.97,4675
fat_content,object,0,0.00,2
shelf_visibility,float64,0,0.00,1732
product_category,object,0,0.00,48
product_price,float64,0,0.00,5818
store_code,object,0,0.00,10
store_age_years,int64,0,0.00,9
store_size,object,1919,28.15,3


In [20]:
#Missing-value Audit
missing_audit = pd.DataFrame(
    {
    "train_missing": train.isna().sum(),
    "train_missing_pct": (train.isna().mean() * 100).round(2),
    "test_missing": test.isna().sum(),
    "test_missing_pct": (test.isna().mean() * 100).round(2)
    }
)

missing_audit = missing_audit.sort_values(
    "train_missing",
    ascending=False
)

missing_audit

,train_missing,train_missing_pct,test_missing,test_missing_pct
store_size,1919,28.15,491.0,28.80
product_weight_kg,1225,17.97,306.0,17.95
fat_content,0,0.00,0.0,0.00
id,0,0.00,0.0,0.00
product_category,0,0.00,0.0,0.00
product_code,0,0.00,0.0,0.00
product_price,0,0.00,0.0,0.00
shelf_visibility,0,0.00,0.0,0.00
store_age_years,0,0.00,0.0,0.00
store_code,0,0.00,0.0,0.00


In [21]:
#Total Missing Values
print("Total missing cells:")

print("Train:", train.isna().sum().sum())
print("Test :", test.isna().sum().sum())

Total missing cells:
Train: 3144
Test : 797


In [22]:
# Unique Value Audit
cardinality = pd.DataFrame(
    {
    "dtype": train.dtypes.astype(str),
    "n_unique": train.nunique(dropna=True),
    "unique_pct": (train.nunique(dropna=True) / len(train) * 100).round(2)
    }
    )

cardinality = cardinality.sort_values("n_unique", ascending = False)

cardinality

,dtype,n_unique,unique_pct
id,object,6818,100.00
total_sales,float64,6776,99.38
product_price,float64,5818,85.33
product_weight_kg,float64,4675,68.57
shelf_visibility,float64,1732,25.40
product_code,object,1555,22.81
product_category,object,48,0.70
store_code,object,10,0.15
store_age_years,int64,9,0.13
store_format,object,4,0.06


In [23]:
# Exact duplicates
print("Exact Duplicate Rows:")

print("Train:", train.duplicated().sum())
print("Test :", test.duplicated().sum())

Exact Duplicate Rows:
Train: 0
Test : 0


In [24]:
# Product Store duplicates
pair_columns = ["product_code", "store_code"]

train_pair_duplicates = train.duplicated(subset=pair_columns).sum()

test_pair_duplicates = test.duplicated(
    subset=pair_columns
).sum()

print("Duplicate Product-Store Combinations")

print("Train:", train_pair_duplicates)
print("Test :", test_pair_duplicates)

Duplicate Product-Store Combinations
Train: 0
Test : 0


In [25]:
# Identify Categorical Columns
categorical_columns = train.select_dtypes(
    include=["object", "category"]
).columns.tolist()

print("Categorical columns:")
print(categorical_columns)

Categorical columns:
['id', 'product_code', 'fat_content', 'product_category', 'store_code', 'store_size', 'store_location_tier', 'store_format']


In [26]:
# Compare Categorical Domains and discover if categories exist in test but not train
for column in categorical_columns:

    train_values = set(train[column].dropna().unique())
    test_values = set(test[column].dropna().unique())

    test_only_values = test_values - train_values

    print(f"\n{column}")
    print("Train unique values:", len(train_values))
    print("Test unique values :", len(test_values))
    print("Test-only values   :", test_only_values)


id
Train unique values: 6818
Test unique values : 1705
Test-only values   : {'row_00825', 'row_01412', 'row_05004', 'row_07534', 'row_07649', 'row_02650', 'row_08479', 'row_02691', 'row_00125', 'row_02344', 'row_03658', 'row_05063', 'row_01573', 'row_02885', 'row_05971', 'row_05360', 'row_02849', 'row_05406', 'row_05246', 'row_07030', 'row_02332', 'row_08481', 'row_02235', 'row_07592', 'row_07375', 'row_03557', 'row_06785', 'row_08031', 'row_07514', 'row_03131', 'row_00559', 'row_00202', 'row_03246', 'row_07380', 'row_08248', 'row_06651', 'row_06528', 'row_03750', 'row_02323', 'row_07996', 'row_03539', 'row_00613', 'row_04306', 'row_03604', 'row_03998', 'row_03233', 'row_04358', 'row_07235', 'row_02238', 'row_01104', 'row_01594', 'row_02940', 'row_05392', 'row_00673', 'row_05806', 'row_06191', 'row_08032', 'row_02079', 'row_00874', 'row_05141', 'row_04333', 'row_08298', 'row_05658', 'row_08029', 'row_00472', 'row_05605', 'row_07075', 'row_05047', 'row_00325', 'row_00804', 'row_01407',

In [27]:
# Target Statistics
target = "total_sales"

print("TARGET:", target)
display(train[target].describe().to_frame().T)

TARGET: total_sales


,count,mean,std,min,25%,50%,75%,max
total_sales,6818.0,2174.756574,1697.894956,32.7,834.7925,1790.89,3092.5875,12996.82


In [28]:
# Target Validity
print("Missing target values :", train[target].isna().sum())
print("Zero target values    :", (train[target] == 0).sum())
print("Negative target values:", (train[target] < 0).sum())
print("Unique target values  :", train[target].nunique())

Missing target values : 0
Zero target values    : 0
Negative target values: 0
Unique target values  : 6776


In [29]:
# Target Quantiles
target_quantiles = train[target].quantile(
    [ 0.00, 0.01, 0.05, 0.10, 0.25, 0.50, 0.75, 0.90, 0.95, 0.99, 1.00]
)

target_quantiles

0.00       32.7000
0.01       81.2017
0.05      188.1190
0.10      342.0420
0.25      834.7925
0.50     1790.8900
0.75     3092.5875
0.90     4557.8410
0.95     5557.1910
0.99     7261.3893
1.00    12996.8200
Name: total_sales, dtype: float64

In [30]:
# Dataset Contract
target = "total_sales"

# Target exists only in train
assert target in train.columns
assert target not in test.columns

# Train and test predictors must match
train_features = set(train.columns) - {target}
test_features = set(test.columns)

assert train_features == test_features

# IDs should be unique
assert train["id"].is_unique
assert test["id"].is_unique

# Target should not contain missing values
assert train[target].notna().all()

print("Target correctly identified.")
print("Train/test predictor columns match.")
print("Train IDs are unique.")
print("Test IDs are unique.")
print("Training target contains no missing values.")
print("\nDataset Contract Passed.")

Target correctly identified.
Train/test predictor columns match.
Train IDs are unique.
Test IDs are unique.
Training target contains no missing values.

Dataset Contract Passed.


## Findings / Inferences

The dataset contains 6,818 training observations and 1,705 test
observations with 12 predictor variables.

Two variables contain missing values:

- `store_size`: 28.15% train, 28.80% test
- `product_weight_kg`: 17.97% train, 17.95% test

The similar missingness rates across train and test suggest no obvious
train/test missingness shift at this stage, although the mechanism of
missingness requires further investigation.

`product_code` has high cardinality (1,555 unique values), while
`store_code` has only 10 unique stores. This suggests that product
identity and store-level effects may be important components of the
prediction problem.

There are no exact duplicate rows and no duplicate
`product_code`-`store_code` combinations within either split.

The target `total_sales` contains no missing, zero, or negative values.
Its distribution is strongly right-skewed, with a mean of 2,174.76,
median of 1,790.89, and maximum of 12,996.82.

Because the competition metric is RMSE, the upper tail of the target
distribution may have a substantial influence on model performance.

No imputation, encoding, transformation, or feature engineering was
performed during this audit.